syne_tune/optimizer/schedulers/hyperband_stopping.py

```diff
@@ -34,58 +34,154 @@ def quantile_cutoff(values, prom_quant, mode):
-        # The data entry in _rungs is a dict mapping task_key to
-        # reward_value
-        assert len(rung_levels) == len(promote_quantiles)
-        self._mode = mode
+    def __init__(self, metric: str, mode: str, resource_attr: str):
         self._metric = metric
+        self._mode = mode
         self._resource_attr = resource_attr
-        self._rungs = [
-            RungEntry(level=x, prom_quant=y, data=dict())
-            for x, y in reversed(list(zip(rung_levels, promote_quantiles)))]
 
-    def on_task_schedule(self):
+    def on_task_schedule(self) -> dict:
+        """
+        Called when new task is to be scheduled. For a promotion-bqsed
+        rung system, check whether any trial can be promoted. If so,
+        return `dict(trial_id, resume_from, milestone)`, where
+        `resume_from` is the rung level where the trial (to be promoted)
+        is paused, and `milestone` is the next rung level it will reach
+        (may be None).
+
+        :return: dict(trial_id, resume_from, milestone) or dict()
+        """
         return dict()
 
-    def on_task_add(self, trial_id, skip_rungs, **kwargs):
+    def on_task_add(self, trial_id: str, skip_rungs: int, **kwargs):
+        """
+        Called when new task is started.
+
+        :param trial_id:
+        :param skip_rungs: This number of lowest rung levels are not
+            considered milestones for this task
+        :param kwargs:
+        """
         pass
 
-    def _cutoff(self, recorded, prom_quant):
-        values = list(recorded.values())
-        return quantile_cutoff(values, prom_quant, self._mode)
+    def on_task_report(
+            self, trial_id: str, result: dict, skip_rungs: int) -> dict:
+        """
+        Called when a trial reports metric results. Returns dict with
+        `milestone_reached` (trial reaches its milestone), `task_continues`
+        (trial should continue; otherwise it is stopped or paused),
+        `next_milestone` (next milestone it will reach, or None).
+        For certain subclasses, there may be additional fields.
+
+        :param trial_id:
+        :param result: Reported metrics
+        :param skip_rungs: This number of lowest rung levels are not
+            considered milestones for this task
+        :return: See above
+        """
+        raise NotImplementedError
 
-    def on_task_report(self, trial_id, result, skip_rungs):
+    def on_task_remove(self, trial_id: str):
         """
-        Decision on whether task may continue (task_continues = True), or
-        should be stopped (task_continues = False).
-        milestone_reached is a flag whether resource coincides with a
-        milestone. If True, next_milestone is the next milestone after
-        resource, or None if there is none.
+        Called when task is removed.
 
         :param trial_id:
-        :param result:
+        """
+        pass
+
+    def get_first_milestone(self, skip_rungs: int) -> int:
```

Comment:
Looks like we can implement this function and reuse in promotion/stopping. Same for other non-implemented functions?